safehttp/plugins/xsrf/xsrf.go

```diff
@@ -0,0 +1,172 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package xsrf
+
+import (
+	"errors"
+	"fmt"
+	"github.com/google/go-safeweb/safehttp"
+	// TODO(@empijei, @kele, @mattiasgrenfeldt, @mihalimara22): decide whether we want to depend on this package or reimplement thefunctionality
+	"golang.org/x/net/xsrftoken"
+)
+
+const (
+	// FMEnforceMode TODO
+	FMEnforceMode = "enforce"
+	// FMReportMode TODO
+	FMReportMode = "report"
+	// FMDisabledMode TODO
+	FMDisabledMode = "disable"
+	// TokenKey is the form key used when sending the token as part of POST
+	// request
+	TokenKey = "xsrf-token"
+)
+
+// StorageService is an interface the framework users need to implement. This
+// will contain information about users of the web application, including their
+// IDs, needed in generating the XSRF token.
+type StorageService interface {
+	// TODO(@mihalimara22): add the parameters that the storage service needs in
+	// order to determine the user ID
+	GetUserID() (string, error)
+}
+
+// Plugin implements XSRF protection.
+type Plugin struct {
+	// TODO(@mihalimara22): decide whether  we want to allow the user to set a
+	// Fetch Metadata policy and how the report mode is going to be handled
+	appKey       string
+	s            StorageService
+	fmPolicyMode string
+	allowedCORS  map[string]bool
+}
+
+// NewPlugin creates a new XSRF plugin with safe defaults
+func NewPlugin(appKey string, s StorageService, domains ...string) *Plugin {
+	// TODO(@mihalimara22): make endpoints a variadic of string literals
+	ad := map[string]bool{}
+	for _, p := range domains {
+		ad[p] = true
+	}
+	return &Plugin{
+		appKey:       appKey,
+		s:            s,
+		fmPolicyMode: FMEnforceMode,
+		allowedCORS:  ad,
+	}
+}
+
+// SetFetchMetadataPolicyMode allows setting the Fetch Metadata policy, enforced
+// by default to one of the other options. Setting it to report will allow
+// requests that violate the policy to pass but will log violation. Setting it
+// to disable
+func (p *Plugin) SetFetchMetadataPolicyMode(mode string) error {
+	switch mode {
+	case FMEnforceMode, FMDisabledMode, FMReportMode:
+		p.fmPolicyMode = mode
+	default:
+		return errors.New("invalid Fetch Metadata policy")
+	}
+	return nil
+}
+
+// GenerateToken generates a cryptographically safe XSRF token using the user ID
+// and the path.
+func (p *Plugin) GenerateToken(host string) (string, error) {
+	userID, err := p.s.GetUserID()
+	if err != nil {
+		return "", fmt.Errorf("token generation failed: %v", err)
+	}
+	token := xsrftoken.Generate(p.appKey, userID, host)
```

Comment:
I think that the `actionID` should be more specific than only the host. I think it should be a combination of host + path + method or at least host + path. That feels more secure.